source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+         "AllOperandsSatisfy() must be called with an OpSpecConstantOp "
+         "instruction");
+  for (uint32_t i = 1; i < inst.NumInOperands(); i++) {
+    uint32_t id = inst.GetSingleWordInOperand(i);
+    if (!f(id)) return false;
+  }
+  return true;
+}
+
+// A helper function to check the type for component wise operations. Returns
+// true if the type:
+//  1) is bool type;
+//  2) is 32-bit int type;
+//  3) is vector of bool type;
+//  4) is vector of 32-bit integer type.
+// Otherwise returns false.
+bool IsValidTypeForComponentWiseOperation(const analysis::Type* type) {
+  if (type->AsBool()) {
+    return true;
+  } else if (auto* it = type->AsInteger()) {
+    if (it->width() == 32) return true;
+  } else if (auto* vt = type->AsVector()) {
+    if (vt->element_type()->AsBool())
+      return true;
+    else if (auto* vit = vt->element_type()->AsInteger()) {
+      if (vit->width() == 32) return true;
+    }
+  }
+  return false;
+}
+}
+
+ir::Instruction* FoldSpecConstantOpAndCompositePass::DoComponentWiseOperation(
+    ir::Module::inst_iterator* pos) {
+  const ir::Instruction* inst = &**pos;
+  const analysis::Type* result_type = GetType(inst);
+  SpvOp spec_opcode = static_cast<SpvOp>(inst->GetSingleWordInOperand(0));
+  // Check and collect operands.
+  std::vector<analysis::Const*> operands;
+  if (!AllOperandsSatisfy(*inst, [&operands, this](uint32_t id) {
+        if (analysis::Const* c = FindRecordedConst(id)) {
+          if (IsValidTypeForComponentWiseOperation(c->type())) {
+            operands.push_back(c);
+            return true;
+          }
+        }
+        return false;
+      })) {
+    return nullptr;
+  }
+
+  if (result_type->AsInteger() || result_type->AsBool()) {
+    // Scalar operation
+    uint32_t result_val = OperateScalars(spec_opcode, operands);
+    auto result_const = CreateConst(result_type, {result_val});
+    return BuildInstructionAndAddToModule(std::move(result_const), pos);
+  } else {
+    // Vector operation
+    const analysis::Type* element_type =
+        result_type->AsVector()->element_type();
+    uint32_t num_dims = result_type->AsVector()->element_count();
+    std::vector<uint32_t> result_vec =
+        OperateVectors(spec_opcode, num_dims, operands);
+    std::vector<const analysis::Const*> result_vector_components;
+    for (uint32_t r : result_vec) {
+      auto rc = CreateConst(element_type, {r});
+      result_vector_components.push_back(rc.get());
+      BuildInstructionAndAddToModule(std::move(rc), pos);
+    }
+    auto new_vec_const = MakeUnique<analysis::VectorConst>(
+        result_type->AsVector(), result_vector_components);
+    return BuildInstructionAndAddToModule(std::move(new_vec_const), pos);
+  }
+}
+
+ir::Instruction*
+FoldSpecConstantOpAndCompositePass::BuildInstructionAndAddToModule(
+    std::unique_ptr<analysis::Const> c, ir::Module::inst_iterator* pos) {
+  analysis::Const* new_const = c.get();
+  uint32_t new_id = ++max_id_;
```

Comment:
Add call to module->SetIdBound() after this line in e332d4d.
Now the pass needs to hold a pointer to the module under process.